### ISSUE TITLE
Permission denied error in Build but not Activate

### ISSUE THREAD

**Issue by Rabadash8820** [webbertakken/unity-actions#26]
I'm not sure if this is user error or a bug. I am trying to create an Android build with Unity 2019.2.17f1, but for some reason the `Unity - Builder` step is failing due to permissions issues, even though `Unity - Activate` and `Unity - Test` both succeed using the same credentials (see screenshot below). I might have typed my password wrong into GitHub secrets, so could it just be that `Build` was the first step to use the password? Or do I need to return a license between `Test` and `Build` perhaps, so that I don't hit the license' max activations? Maybe I can just skip the `Build` step entirely, and output the artifacts created by `Test`?

Screenshot:

![image](https://user-images.githubusercontent.com/8388785/71935758-a3f68c00-3175-11ea-8783-9bed8570e050.png)

Here is my config:

```yml
name: Unity and Unit Tests

on:
  pull_request:
    branches:
      - master

jobs:
  unity:
    runs-on: ubuntu-latest
    strategy:
      max-parallel: 1   # Only let one job run at a time, since we can only have 2 activations of Unity license key (including user)
      fail-fast: true   # Cancel all parallel jobs if one of them fails. Technically not necessary with max-parallel=1, but good to keep around as a reminder
      matrix:
        targetPlatform: [ Android, iOS ]
    steps:
      - name: Checkout
        uses: actions/checkout@v2.0.0
        with:
          lfs: true
      - name: Unity - Activate
        uses: webbertakken/unity-activate@v1
        env:
          UNITY_EMAIL: ${{ secrets.UNITY_EMAIL }}
          UNITY_PASSWORD: ${{ secrets.UNITY_PASSWORD }}
          UNITY_SERIAL: ${{ secrets.UNITY_SERIAL }}
      - name: Unity - Test
        uses: webbertakken/unity-test-runner@v1.1.1
        env:
          TEST_MODE: all
          PROJECT_PATH: HighHandHoldem
      - name: Unity - Build
        uses: webbertakken/unity-builder@v0.4
        with:
          unityVersion: 2019.2.17f1
          targetPlatform: ${{ matrix.targetPlatform }}
          projectPath: HighHandHoldem
      - name: Unity - Return License
        uses: webbertakken/unity-return-license@v1
```

Here is what seems to be the relevant portion of the `Build` logs:

```txt
###########################
#    Building platform    #
###########################

Initiating legacy licensing module

LICENSE SYSTEM [202017 16:23:5] Next license update check is after 2020-01-08T21:14:27


LICENSE SYSTEM [202017 16:23:5] 9d0d7fef574d4a158d8ad4919b973caf != d39b8e2f4d364b2e98b06afa0c6e08c5


LICENSE SYSTEM [202017 16:23:5] 9d0d7fef574d4a158d8ad4919b973caf != d39b8e2f4d364b2e98b06afa0c6e08c5

DisplayProgressbar: Unity license
[UnityConnectServicesConfig] config is NOT valid, switching to default
[0107/162306:ERROR:browser_main_loop.cc(161)] Running without the SUID sandbox! See https://code.google.com/p/chromium/wiki/LinuxSUIDSandboxDevelopment for more information on developing with the sandbox on.
Home directory not accessible: Permission denied
[0107/162306:ERROR:sandbox_linux.cc(308)] InitializeSandbox() called with multiple threads in process gpu-process
Cancelling DisplayDialog: Failed to activate/update license Missing or bad username or password. Please try again using valid credentials or contact support@unity3d.com
This should not be called in batch mode.
 
(Filename: ./Editor/Platform/Linux/EditorUtility.cpp Line: 274)

Build failed, with exit code 1
```

**Comment by thorgeirk11** [comment]
I am running into the same issue

**Comment by webbertakken** [comment]
Thanks for bringing up this issue!

This is probably a bug. This is the kind of error that typically occurs if you activate on a different version than you run the build on.

I won't be able to fix it in the coming week but I will have a look at it after. Also note that I'm moving licensing back in to this action itself in https://github.com/webbertakken/unity-builder/pull/12

In the meantime feel free to use earlier versions of the action or send a PR if you know what causes the issue.

**Comment by thorgeirk11** [comment]
Thanks for the quick reply!

**Comment by DanVicarel** [comment]
@webbertakken Thanks for the info. How will that PR affect the Test action? If I want to run Test before Build, would I still need to use the separate Activate action, or would you now recommend running Test afterward?

**Comment by webbertakken** [comment]
> @webbertakken Thanks for the info. How will that PR affect the Test action? If I want to run Test before Build, would I still need to use the separate Activate action, or would you now recommend running Test afterward?

Hi @DanVicarel, excellent question. Currently you would still need to run the separate Activate action, but the idea is indeed to make each action independent. So the goal is to essentially do the same for Test runner as in the above mentioned PR for Builder. There is an issue about it here: https://github.com/webbertakken/unity-test-runner/issues/9

**Comment by webbertakken** [comment]
Could you please try if https://github.com/webbertakken/unity-builder/releases/tag/v0.5 fixes this problem for you?

**Comment by Rabadash8820** [comment]
So to be clear, running Test before Build would look something like this:

```yml
- Activate
- Test
- Return License
- Build
```

at least until licensing is integrated with Test?

(Btw, I am both @Rabadash8820 and @DanVicarel; one is a personal account and one is an old work account, and I haven't settled on one yet 😆)

**Comment by Rabadash8820** [comment]
Oh, just had a closer look at the [`main.yml`](https://github.com/webbertakken/unity-actions/blob/master/.github/workflows/main.yml) example in this repo. It never occurred to me to do testing and building in separate jobs. That actually makes way more sense, so then edit mode tests, play mode tests, and a matrix of platform-specific builds can all run in parallel. The only limiting factor is my Unity Plus license, which can only be activated on two machines at a time, and one of them is for my PC 😞 Still, good approach.

**Comment by Rabadash8820** [comment]
@webbertakken Unfortunately, I'm still getting Permission denied errors with v0.5. If it helps, I could grant you read-only permissions on my private repo so you can see my `main.yml` and Action logs. My situation is still the same as above though... Activate/Test/Return steps succeed, but Build step fails with following log:

```txt
Run webbertakken/unity-builder@v0.5
  with:
    unityVersion: 2019.2.17f1
    targetPlatform: Android
    projectPath: HighHandHoldem
  env:
    UNITY_EMAIL: ***
    UNITY_PASSWORD: ***
    UNITY_SERIAL: ***
/usr/bin/docker build /home/runner/work/_actions/webbertakken/unity-builder/v0.5/builder --file /home/runner/work/_actions/webbertakken/unity-builder/v0.5/builder/Dockerfile --build-arg IMAGE=gableroux/unity3d:2019.2.17f1-android --tag unity-builder:2019.2.17f1-android
Sending build context to Docker daemon  260.1kB

Step 1/15 : ARG IMAGE
Step 2/15 : FROM $IMAGE
2019.2.17f1-android: Pulling from gableroux/unity3d
2746a4a261c9: Pulling fs layer
...
...
...
Successfully built 7305a4d97685
Successfully tagged unity-builder:2019.2.17f1-android
/usr/bin/docker run --workdir /github/workspace --rm --env UNITY_LICENSE --env UNITY_VERSION=2019.2.17f1 --env PROJECT_PATH=HighHandHoldem --env BUILD_TARGET=Android --env BUILD_NAME= --env BUILDS_PATH= --env BUILD_METHOD= --env HOME=/github/home --env GITHUB_REF --env GITHUB_SHA --env GITHUB_REPOSITORY --env GITHUB_ACTOR --env GITHUB_WORKFLOW --env GITHUB_HEAD_REF --env GITHUB_BASE_REF --env GITHUB_EVENT_NAME --env GITHUB_WORKSPACE=/github/workspace --env GITHUB_ACTION --env GITHUB_EVENT_PATH --env RUNNER_OS --env RUNNER_TOOL_CACHE --env RUNNER_TEMP --env RUNNER_WORKSPACE --volume /var/run/docker.sock:/var/run/docker.sock --volume /home/runner/work/_temp/_github_home:/github/home --volume /home/runner/work/_temp/_github_workflow:/github/workflow --volume /home/runner/work/HighHandHoldem/HighHandHoldem:/github/workspace unity-builder:2019.2.17f1-android
Initiating legacy licensing module
DisplayProgressbar: Unity license
[0111/155243:ERROR:browser_main_loop.cc(161)] Running without the SUID sandbox! See https://code.google.com/p/chromium/wiki/LinuxSUIDSandboxDevelopment for more information on developing with the sandbox on.
[0111/155243:ERROR:sandbox_linux.cc(308)] InitializeSandbox() called with multiple threads in process gpu-process
[UnityConnectServicesConfig] config is NOT valid, switching to default
UnityConnectLoginRequest: Failed to login - please check your username or password
 
(Filename: ./Editor/Src/UnityConnect/UnityConnectLoginRequest.cpp Line: 31)

No sufficient permissions while processing request "https://core.cloud.unity3d.com/api/login", HTTP error code 401
.Cancelling DisplayDialog: Failed to activate/update license Timeout occured while trying to update license. Please try again later or contact support@unity3d.com
This should not be called in batch mode.
 
(Filename: ./Editor/Platform/Linux/EditorUtility.cpp Line: 274)

Unclassified error occured while trying to activate (professional) license.
Exit code was: 1
##[error]The process '/usr/bin/docker' failed with exit code 1
##[error]Node run failed with exit code 1

```

**Comment by webbertakken** [comment]
@Rabadash8820 I'm not near a pc the coming week but from what I can see from my phone is that in `src/model/docker.js` the variables for Pro license are never passed to the docker container, like `UNITY_LICENSE` is.

I can fix it later when I have access to a pc again. In the meantime you could try with a free license or make a PR to add the variables `UNITY_EMAIL`, `UNITY_PASSWORD`, `UNITY_SERIAL` to the run command in docker.js. They can be passed the same way as `UNITY_LICENSE`.

**Comment by webbertakken** [comment]
Could you try again with [builder v0.6](https://github.com/webbertakken/unity-builder/releases/tag/v0.6)?

**Comment by webbertakken** [comment]
Yes. Since v0.5 licensing is integrated with builder. You should provide `UNITY_LICENSE` as an env variable to the builder action as per the updated readme. Reason is exactly that; different versions require different license activations. 

Let me know if that fixes your problem. If it doesn't we should create a new ticket for each specific version this occurs for.

**Comment by thorgeirk11** [comment]
Ah, thanks! Will try that out

**Comment by webbertakken** [comment]
> Ah, thanks! Will try that out

@thorgeirk11 did it work for you?

**Comment by thorgeirk11** [comment]
Did not manage to get it to work yet 
```
      - name: Make Unity Client
        uses: webbertakken/unity-builder@v0.6
        env:
          UNITY_EMAIL:    ${{ secrets.UNITY_EMAIL }}
          UNITY_PASSWORD: ${{ secrets.UNITY_PASSWORD }}
          UNITY_SERIAL:   ${{ secrets.UNITY_SERIAL }}
        with:
          unityVersion: 2018.4.10f1
          targetPlatform: StandaloneWindows64
          projectPath: Client/
```

Just to be clear ... The serial should look like SC-XXXX-XXXX-XXXX-XXXX-XXXX ?

**Comment by webbertakken** [comment]
@thorgeirk11 I have never seen a pro license but I would guess that is indeed what it would look like.

Could you try the same workflow with unity version 2019.2.11f1 to verify your credentials are correct? And if they are, we need to create a new issue for activating in 2018 versions, as I've never tested any.

**Comment by Jibbajabbafic** [comment]
I think it would be useful to update the 'Request Activation File' action to allow you to pass the unity version like in the Build action so licenses for other version can easily be generated.

**Comment by DanVicarel** [comment]
Personally I dont really see the value in that Action. There are still several manual steps involved, and they're already [documented](https://docs.unity3d.com/Manual/CommandLineArguments.html) in the Unity Manual. Plus, users already have the version of Unity installed that they would want to use, so they would get the correct license by running the command themselves. Seems like u could just direct users to those docs, and then the "first" action of yours that they use would be Activate (or one of the Actions that has activation built in).

**Comment by Jibbajabbafic** [comment]
@DanVicarel The problem is that the license file is for a specific machine, so you can't generate and active one on your own computer and then use it for GH actions.

**Comment by Rabadash8820** [comment]
@Jibbajabbafic Ah, that does make sense. I stand corrected. In that case yeah it would be handy to pass a Unity version to that action.

**Comment by webbertakken** [comment]
@Rabadash8820 this should now work on the master version of builder. Could you give it a go?

**Comment by Rabadash8820** [comment]
@webbertakken License activation/return appears to be working fine. The build is still failing, but I assume that's because I'm trying to build for Android, which it looks like you don't officially support yet? I see the following logs, which may be culprit:

```txt
DisplayProgressbar: Detect Android NDK
Uploading Crash Report
UnityException: Android NDK not found
Unable to locate Android NDK.
```

Anyway, I think Android-specific errors are outside the scope of this Issue.

**Comment by webbertakken** [comment]
Thanks for verifying!

Support for all platforms is tracked here https://github.com/webbertakken/unity-actions/issues/12

Closing this issue as the bug was resolved.
